addr_manager_ds.go

```diff
@@ -0,0 +1,316 @@
+	prefix := ds.NewKey(p.Pretty())
+	q := query.Query{Prefix: prefix.String()}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []ma.Multiaddr{}
+	}
+
+	var addrs []ma.Multiaddr
+	for result := range results.Next() {
+		addrbytes := result.Value.([]byte)
+		addr, err := ma.NewMultiaddrBytes(addrbytes)
+		if err != nil {
+			continue
+		}
+		addrs = append(addrs, addr)
+	}
+
+	return addrs
+}
+
+// Peers returns all of the peer IDs for which the AddrBook has addresses.
+func (mgr *DatastoreAddrManager) Peers() []peer.ID {
+	q := query.Query{}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []peer.ID{}
+	}
+
+	idset := make(map[peer.ID]struct{})
+	for result := range results.Next() {
+		key := ds.RawKey(result.Key)
+		id, err := peerIDFromKey(key)
+		if err != nil {
+			continue
+		}
+		idset[id] = struct{}{}
+	}
+
+	ids := make([]peer.ID, 0, len(idset))
+	for id := range idset {
+		ids = append(ids, id)
+	}
+	return ids
+}
+
+// AddrStream returns a channel on which all new addresses discovered for a
+// given peer ID will be published.
+func (mgr *DatastoreAddrManager) AddrStream(ctx context.Context, p peer.ID) <-chan ma.Multiaddr {
+	initial := mgr.Addrs(p)
+	return mgr.subsManager.AddrStream(ctx, p, initial)
+}
+
+// ClearAddrs will delete all known addresses for a peer ID.
+func (mgr *DatastoreAddrManager) ClearAddrs(p peer.ID) {
+	prefix := ds.NewKey(p.Pretty())
+	for i := 0; i < dsWriteRetries; i++ {
+		q := query.Query{Prefix: prefix.String()}
+		results, err := mgr.ds.Query(q)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		batch, err := mgr.ds.Batch()
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		for result := range results.Next() {
+			batch.Delete(ds.NewKey(result.Key))
+		}
+		if err = batch.Commit(); err != nil {
+			log.Errorf("failed to clear addresses for peer %s: %s\n", p.Pretty(), err)
+			continue
+		}
+		mgr.ttlManager.clear(ds.NewKey(p.Pretty()))
+		return
+	}
+	log.Errorf("failed to clear addresses for peer %s after %d attempts\n", p.Pretty(), dsWriteRetries)
+}
+
+// ttlmanager
+
+type ttlentry struct {
+	TTL       time.Duration
+	ExpiresAt time.Time
+}
+
+type ttlmanager struct {
+	sync.RWMutex
+	entries map[ds.Key]*ttlentry
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	ticker *time.Ticker
+	ds     ds.Datastore
+}
+
+func newTTLManager(parent context.Context, d ds.Datastore, tick time.Duration) *ttlmanager {
+	ctx, cancel := context.WithCancel(parent)
+	mgr := &ttlmanager{
+		entries: make(map[ds.Key]*ttlentry),
+		ctx:     ctx,
+		cancel:  cancel,
+		ticker:  time.NewTicker(tick),
+		ds:      d,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-mgr.ctx.Done():
+				mgr.ticker.Stop()
+				return
+			case <-mgr.ticker.C:
+				mgr.tick()
+			}
+		}
+	}()
+
+	return mgr
+}
+
+// To be called by TTL manager's coroutine only.
+func (mgr *ttlmanager) tick() {
+	mgr.RLock()
```

Comment:
this can't just be an `RLock`, delete is a writing operation.

Would be great to write a test, that when run with `go test -race` fails due to this race condition.

Reply:
another great catch, thanks! and yeah i need to look into the `-race` testing